test/e2e/xfn_test.go

```diff
@@ -0,0 +1,320 @@
+		caPem, keyPem, err := createCert(dnsName)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "reg-cert",
+				Namespace: ns,
+			},
+			Type: corev1.SecretTypeTLS,
+			StringData: map[string]string{
+				"tls.crt": caPem,
+				"tls.key": keyPem,
+			},
+		}
+		client := config.Client().Resources()
+		if err := client.Create(ctx, secret); err != nil {
+			t.Fatalf("Cannot create secret %s: %v", secret.Name, err)
+		}
+		configMap := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "reg-ca",
+				Namespace: namespace,
+			},
+			Data: map[string]string{
+				"domain.crt": caPem,
+			},
+		}
+		if err := client.Create(ctx, configMap); err != nil {
+			t.Fatalf("Cannot create config %s: %v", configMap.Name, err)
+		}
+		return ctx
+	}
+}
+
+// CopyImagesToRegistry copies fn images to private registry
+func CopyImagesToRegistry() features.Func {
+	return func(ctx context.Context, t *testing.T, config *envconf.Config) context.Context {
+		nodes := &corev1.NodeList{}
+		if err := config.Client().Resources().List(ctx, nodes); err != nil {
+			t.Fatal("cannot list nodes", err)
+		}
+		if len(nodes.Items) == 0 {
+			t.Fatalf("no nodes in the cluster")
+		}
+		i := slices.IndexFunc(nodes.Items[0].Status.Addresses, func(a corev1.NodeAddress) bool {
+			return a.Type == corev1.NodeInternalIP
+		})
+		if i == -1 {
+			t.Fatalf("no nodes with private address")
+		}
+		addr := nodes.Items[0].Status.Addresses[i].Address
+		p := gexe.RunProc(fmt.Sprintf("skopeo copy docker-daemon:crossplane-e2e/fn-labelizer:latest docker://%s:32000/fn-labelizer:latest --dest-tls-verify=false", addr)).Wait()
+		out, _ := io.ReadAll(p.Out())
+		t.Logf("skopeo stdout: %s", string(out))
+		if p.ExitCode() != 0 {
+			t.Fatalf("copying image to registry not successful, exit code %v std out %v std err %v", p.ExitCode(), string(out), p.Err())
+		}
+		return ctx
+	}
+}
+
+// InstallDockerRegistry with custom TLS
+func InstallDockerRegistry() features.Func {
+	return funcs.AsFeaturesFunc(funcs.HelmInstall(
+		helm.WithName("private"),
+		helm.WithNamespace("reg"),
+		helm.WithWait(),
+		helm.WithChart("https://helm.twun.io/docker-registry-2.2.1.tgz"),
+		helm.WithArgs(
+			"--set service.type=NodePort",
+			"--set service.nodePort=32000",
+			"--set tlsSecretName=reg-cert",
+		),
+	))
+}
+
+// CrossplaneDeployedWithFunctionsEnabled asserts that crossplane deployment with composition functions is enabled
+func CrossplaneDeployedWithFunctionsEnabled() features.Func {
+	return funcs.AllOf(
+		funcs.AsFeaturesFunc(funcs.HelmUpgrade(
+			HelmOptions(
+				helm.WithArgs(
+					"--set args={--debug,--enable-composition-functions}",
+					"--set xfn.enabled=true",
+					"--set xfn.args={--debug}",
+					"--set registryCaBundleConfig.name=reg-ca",
+					"--set registryCaBundleConfig.key=domain.crt",
+					"--set xfn.resources.requests.cpu=100m",
```

Comment:
The e2e test was passing on my machine, but once the PR got created, the test was continuously failing on Github runner. After some debugging, I could see that the root cause was the inability to deploy crossplane with xfn enabled,  saying that none of nodes can fulfill the requested cpu resources. Further digging led me to the default resource requests and limit for `crossplane-xfn` container:

```yaml
  resources:
    limits:
      # -- CPU resource limits for the Composite function runner container.
      cpu: 2000m
      # -- Memory resource limits for the Composite function runner container.
      memory: 2Gi
    requests:
      # -- CPU resource requests for the Composite function runner container.
      cpu: 1000m
      # -- Memory resource requests for the Composite function runner container.
      memory: 1Gi

```

These are too high for Github runner (2 cpus). After lowering cpu requests at Helm chart upgrade, everything started to work again.

The default values got introduced in https://github.com/crossplane/crossplane/commit/0e210698c72fdd1b5cef16718a145d7eb09eeb08 by @negz as a part of https://github.com/crossplane/crossplane/pull/3465 - I could not find any details on the set resource values in that PR, or [the design document](https://github.com/crossplane/crossplane/blob/b9e4773dc431d974a44232b95500e6f815a9b8ac/design/design-doc-composition-functions.md) but I assume that there was a good reason for choosing them @negz ?

Reply:
@pedjak I believe the rationale was that we'd want enough CPU/compute for folks to be able to run at least a small handful of containers at once. For E2E tests I think lower limits/requests should be fine for the time being. (As you know) I'm planning to open an updated Functions design that should make this much less of an issue.